FEA Add `ArccosDistance` (continued)

jgonggrijp opened the pull request:
#### Reference Issues/PRs

This is the continuation of #7829. I worked on this about a year ago and remember that it worked and I was (nearly) ready to submit a PR, but I was too busy at work to actually do it. Now, @ArturoAmorQ seems to plan to redo all of my work in #24099, which prompted me to just submit what I have.

#### What does this implement/fix? Explain your changes.

As originally described by @lmcinnes in #7829:

> Many algorithms, such as word2vec result in nearest neighbor computations based on cosine similarity. Unfortunately, since cosine (dis)similarity is not a metric it can't be used with kd-trees and ball-trees. This means that algorithms that make use of these structures (e.g. DBSCAN clustering, fast t-SNE, etc.) can't operate with regard to the "appropriate" (dis)similarity measure. Here we add angular (or arccos) distance which is the natural metric analogue of cosine dissimilarity to the valid metrics used for kd-trees and ball-trees. Credit for this work belongs to @brunoalano who submitted a similar change to hdbscan.

#### Any other comments?

Due to the urgency, I have not checked whether the branch ticks all the marks; I will correct as needed.

@ArturoAmorQ, please have a look at this branch before investing more effort in #24099.

lmcinnes committed:
Add Angular distance, a metric version of cosine distance that can be used in kd-trees and ball-trees.

lmcinnes committed:
Fix a typo as per suggestion from @rth.

lmcinnes committed:
Make Arccos an allowed metric for ball-trees. (Use in kd-trees still needs to be verified).

lmcinnes committed:
Add arccos distance to tests for ball tree and nearest neighbors.

lmcinnes committed:
Missed a comma.

lmcinnes committed:
Arccos not valid here as we also do brute force search.

lmcinnes committed:
Cosine dissimilarity and arccos are different incomparable metrics.

lmcinnes committed:
cosine not valid for 'auto'

lmcinnes committed:
cosine not valid for 'auto'

jgonggrijp committed:
Merge lmcinnes/arccos_distance_for_trees into angular-distance

jgonggrijp committed:
Test the arccos metric by itself and test it in more places (#7829)

jgonggrijp committed:
Adjust neighbors:test_kde_sample_weights for arccos distance (#7829)

In one-dimensional data, the assertions fail because NaN values appear
in different places. Skipping arccos for d==1 seems defensible,
because the concept of angles degenerates at one dimension.

In ten-dimensional data, the effect of the sample weights was just
under the original magic number of 0.001, at 0.0008073349365247395. As
the original threshold was an undocumented number, I presumed that it
was somewhat arbitrary, which makes lowering a bit seem defensible. I
fully realize, however, that I might be wrong about this.

jgonggrijp committed:
Uncomment integration branch name in benchmark configuration (#7829)

jgonggrijp committed:
Tweak neighbors benchmark temporarily to measure arccos (#7829)

This commit will be reverted.

jgonggrijp committed:
Use CBLAS ddot to implement the arccos distance (#7829)

Gratefully following the hints by @rth and @jnothman.

jgonggrijp committed:
Optimize arccos distance by adding a reduced distance measure (#7829)

jgonggrijp committed:
Make the arccos rdist safer (#7829)

Commit:
Merge branch 'main' into arccos_distance_for_trees

Commit:
fixup! Merge branch 'main' into arccos_distance_for_trees

ArturoAmorQ commented:
I think our work may be complementary, as the only thing I've done so far in https://github.com/scikit-learn/scikit-learn/pull/24099 is to fix the conflicts with `main`. Feel free to fetch the last 2 commits ([`1cc97e9` (#24099)](https://github.com/scikit-learn/scikit-learn/pull/24099/commits/1cc97e9c471fd4aee37878261799dfde05b963bc) and [`c03bc77` (#24099)](https://github.com/scikit-learn/scikit-learn/pull/24099/commits/c03bc77202b299f529d63a1df350cef06b0f30a4)) or use them for inspiration on how to fix the conflicts in this PR.

jgonggrijp commented:
@ArturoAmorQ Thanks!

jgonggrijp committed:
Revert "Tweak neighbors benchmark temporarily to measure arccos (#7829)"

This reverts commit b8837a813ac622b91bc5b6669303c04f07606f58.

jgonggrijp committed:
Merge remote-tracking branch 'ArturoAmorQ/arccos_distance_for_trees' into angular-distance

This brings the work up-to-date with the main branch, moving added
code to new locations.
Ref #7829

jjerphan commented on sklearn/neighbors/tests/test_kde.py:
```diff
@@ -164,8 +171,12 @@ def test_kde_sample_weights():
         for algorithm in ["auto", "ball_tree", "kd_tree"]:
-            for metric in ["euclidean", "minkowski", "manhattan", "chebyshev"]:
-                if algorithm != "kd_tree" or metric in KDTree.valid_metrics:
+            for metric in [
+                "euclidean", "minkowski", "manhattan", "chebyshev", "arccos"
+            ]:
+                if ((metric != "arccos" or d != 1) and
+                    (algorithm != "kd_tree" or metric in KDTree.valid_metrics)
+                ):
```
Is it possible to parametrize `test_kde_sample_weights` on `metric` and `metric` using the `pytest.mark.parametrize` decorator?

jgonggrijp replied:
Looks to me like that should be possible, yes. You wrote "`metric` and `metric`", but I suppose you meant `algorithm` and `metric`? Shall I parametrize `d` as well, while I'm at it?

jjerphan replied:
Yes, that's a typo from me: I meant `algorithm` and `metric`, which can be parametrised with `d` like you propose.

You can also use `pytest.skip` for the combinations of the parameters which aren't meeting the conditions of the test L117 to line 179.

jjerphan commented on sklearn/neighbors/tests/test_kde.py:
```diff
@@ -190,7 +201,7 @@ def test_kde_sample_weights():
                     # Test that sample weights has a non-trivial effect
                     diff = np.max(np.abs(scores_no_weight - scores_weight))
-                    assert diff > 0.001
+                    assert diff > 0.0005
```
What is the reason to change this value?

jgonggrijp replied:
In ten-dimensional data, the effect of the sample weights was just under the original magic number of 0.001, at 0.0008073349365247395. As the original threshold was an undocumented number, I presumed that it was somewhat arbitrary, which makes lowering a bit seem defensible. I fully realize, however, that I might be wrong about this.

(Copied from my commit message in 09eee17b6c6f0.)

jjerphan commented on sklearn/metrics/tests/test_dist_metrics.py:
```diff
@@ -341,6 +341,26 @@ def haversine_slow(x1, x2):
+        for j, x2 in enumerate(a):
+            c[i, j] = arccos_slow(x1, x2)
+
+    assert_array_almost_equal(b, c)
```
```suggestion
    assert_allclose(b, c)
```

jjerphan commented on sklearn/neighbors/_ball_tree.pyx:
```diff
@@ -14,7 +14,8 @@ VALID_METRICS = ['EuclideanDistance', 'SEuclideanDistance',
                  'RogersTanimotoDistance', 'RussellRaoDistance',
                  'SokalMichenerDistance', 'SokalSneathDistance',
-                 'PyFuncDistance', 'HaversineDistance']
+                 'ArccosDistance', 'HaversineDistance',
```
To match the order of `DistanceMetric`'s definition in `sklearn/metrics/_dist_metrics.pyx.tp`
```suggestion
                 'HaversineDistance', 'ArccosDistance',
```

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -2785,6 +2794,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+# most suitable when you wish to use cosine distance.
+
+cdef class ArccosDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+    cdef inline {{INPUT_DTYPE_t}} cosine_similarity(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        cdef {{INPUT_DTYPE_t}} d = ddot(size, x1, 1, x2, 1)
+        cdef {{INPUT_DTYPE_t}} norm1 = ddot(size, x1, 1, x1, 1)
+        cdef {{INPUT_DTYPE_t}} norm2 = ddot(size, x2, 1, x2, 1)
+        return d / sqrt(norm1 * norm2)
+
+    cdef inline {{INPUT_DTYPE_t}} rdist(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        # The non-metric, but rank-preserving, cosine distance.
+        return 1 - self.cosine_similarity(x1, x2, size)
+
+    cdef inline {{INPUT_DTYPE_t}} dist(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        return acos(self.cosine_similarity(x1, x2, size)) / M_PI
```
This:
 - `const` qualify `x1` and `x2` to be able to work with read-only datasets: this is also good practice if the underlying data is only peaked and is not modified for compiler optimisation and for communication. 
 - Return `DTYPE_t` instead of `INPUT_DTYPE_t`: we want to use double precision in all cases. 
```suggestion
    cdef inline {{INPUT_DTYPE_t}} cosine_similarity(
        self,
        const {{INPUT_DTYPE_t}}* x1,
        const {{INPUT_DTYPE_t}}* x2,
        ITYPE_t size
    ) nogil except -1:
        cdef DTYPE_t d = ddot(size, x1, 1, x2, 1)
        cdef DTYPE_t norm1 = ddot(size, x1, 1, x1, 1)
        cdef DTYPE_t norm2 = ddot(size, x2, 1, x2, 1)
        return d / sqrt(norm1 * norm2)

    cdef inline {{INPUT_DTYPE_t}} rdist(
        self,
        const {{INPUT_DTYPE_t}}* x1,
        const {{INPUT_DTYPE_t}}* x2,
        ITYPE_t size
    ) nogil except -1:
        # The non-metric, but rank-preserving, cosine distance.
        return 1 - self.cosine_similarity(x1, x2, size)

    cdef inline {{INPUT_DTYPE_t}} dist(
        self,
        const {{INPUT_DTYPE_t}}* x1,
        const {{INPUT_DTYPE_t}}* x2,
        ITYPE_t size
    ) nogil except -1:
        return acos(self.cosine_similarity(x1, x2, size)) / M_PI
```

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -2785,6 +2794,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+    ) nogil except -1:
+        return acos(self.cosine_similarity(x1, x2, size)) / M_PI
+
+    cdef inline {{INPUT_DTYPE_t}} _rdist_to_dist(
+        self,
+        {{INPUT_DTYPE_t}} rdist
+    ) nogil except -1:
+        return acos(1 - rdist) / M_PI
+
+    cdef inline {{INPUT_DTYPE_t}} _dist_to_rdist(
+        self,
+        {{INPUT_DTYPE_t}} dist
+    ) nogil except -1:
+        return 1 - cos(M_PI * dist)
```
Follow-up of the previous suggestion to return `DTYPE_t` instead of `INPUT_DTYPE_t`.

```suggestion
    cdef inline DTYPE_t _rdist_to_dist(
        self,
        DTYPE_t rdist
    ) nogil except -1:
        return acos(1 - rdist) / M_PI

    cdef inline DTYPE_t _dist_to_rdist(
        self,
        DTYPE_t dist
    ) nogil except -1:
        return 1 - cos(M_PI * dist)
```

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -2785,6 +2794,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        cdef {{INPUT_DTYPE_t}} d = ddot(size, x1, 1, x2, 1)
+        cdef {{INPUT_DTYPE_t}} norm1 = ddot(size, x1, 1, x1, 1)
+        cdef {{INPUT_DTYPE_t}} norm2 = ddot(size, x2, 1, x2, 1)
```
Can you provide benchmarks for between using BLAS `ddot` and manually performing the dot products, as propose in https://github.com/scikit-learn/scikit-learn/pull/7829#discussion_r87914940?

jgonggrijp replied:
I created a makeshift benchmark in b8837a813 by temporarily adjusting another benchmark, which enabled me to test the performance effects of the subsequent three commits (bd22d28e7, f56fab4f7 and 4deba5b21). I then reverted the makeshift benchmark again in 856342319. Please suggest a neat, permanent way to benchmark this; I am pretty much an outsider.

(Since this was a year ago, I don't remember what came out, although I can probably look it up as I remember that the benchmark suite stores results to disk. In any case, I tend to revert commits that complicate the code without improving performance, so there was probably a benefit.)

jjerphan replied:
Hi @jgonggrijp, I am catching up with others topics but I will come back to you shortly. Thanks for your patience.

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -42,7 +42,13 @@ cdef extern from "arrayobject.h":
 
 
-from libc.math cimport fabs, sqrt, exp, pow, cos, sin, asin
+# ddot from CBLAS (cf sklearn/linear_model/cd_fast.pyx:102-117)
+cdef extern from "cblas.h":
+    double ddot "cblas_ddot"(int N, double *X, int incX, double *Y, int incY
+                             ) nogil
```
Internally, we define and uses interfaces for BLAS' which handle various precision using  Cython fused-types. For instance, we have a `_dot` interface for the `sdot` and `ddot` BLAS' subroutines:

https://github.com/scikit-learn/scikit-learn/blob/b571d64066dd13b6f9b5eae399c6a20db39e5a83/sklearn/utils/_cython_blas.pyx#L20-L26

`_dot` must be imported and used inplace of `ddot`.

The import, that must be placed at l.48 of the original file, is:  
```
from ..utils._cython_blas cimport _gemm
```

jgonggrijp replied:
In the above line, I replaced `_gemm` by `_dot`:

```cython
from ..utils._cython_blas cimport _dot
```

and then inserted that at line 50, after removing the former `cdef extern` on line 46-48.

jjerphan replied:
Good catch!

jjerphan reviewed (commented):
Thank you for pursuing this work, @jgonggrijp and @ArturoAmorQ.

Here are a few comments.

jgonggrijp committed:
Use _dot instead of ddot in sklearn/metrics/_dist_metrics (#24106)

As suggested in review comment by @jjerphan
https://github.com/scikit-learn/scikit-learn/pull/24106#discussion_r940049986

jgonggrijp committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -117,6 +118,8 @@ METRIC_MAPPING{{name_suffix}} = {
     'sokalmichener': SokalMichenerDistance{{name_suffix}},
     'sokalsneath': SokalSneathDistance{{name_suffix}},
     'haversine': HaversineDistance{{name_suffix}},
+    'arccos': ArccosDistance{{name_suffix}},
+    'angular': ArccosDistance{{name_suffix}},
```
I think it is better to have only one specification for this distance

I find that `AngularDistance` is a better naming than `ArccosDistance` as it caries more meaning. Hence I would rename this `DistanceMetric` to `AngularDistance` and here have:

```suggestion
    'angular': AngularDistance{{name_suffix}},
```

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -2785,6 +2789,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+# most suitable when you wish to use cosine distance.
+
+cdef class ArccosDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+    cdef inline {{INPUT_DTYPE_t}} cosine_similarity(
+        self,
+        const {{INPUT_DTYPE_t}}* x1,
+        const {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        cdef DTYPE_t d = _dot(size, x1, 1, x2, 1)
+        cdef DTYPE_t norm1 = _dot(size, x1, 1, x1, 1)
+        cdef DTYPE_t norm2 = _dot(size, x2, 1, x2, 1)
+        return d / sqrt(norm1 * norm2)
```
To answer your comment regarding benchmarking various versions of this method (i.e. https://github.com/scikit-learn/scikit-learn/pull/24106#discussion_r944395136), I think the best is to perform a standalone comparison without scikit-learn:

To do this, I recommend that you use:
 - [Cython in a Notebook using annotation mode](https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook) to understand how to minimize the overhead of CPython (there must be some in `def` functions but none in `cdef` function)
 - several `cdef` functions for variations of `cosine_similarity` which define the core of the computations
 - a Python `def` function as a front-end for each `cdef` function. Such function must just transfer the `x1` and `x2` numpy arrays' buffers' pointers to those `cdef` functions.
 - the `%timeit` magic on those Python `def` functions to assert performance of `cdef` functions. You better use large `x1` and `x2` arrays not to have the interaction with CPython have the runtime overhead majority.
 - report the notebook somewhere publicly (e.g. a public repository on GitHub) and link this notebook in a comment here with conclusions. This way reviewers can rerun it and better suggestion on top of your work.

jjerphan reviewed (commented):
After an eternity, I am coming back to reviewing this PR.

Thanks for your patience, @eggie5!
